### DISCUSSION TITLE
'Compute Strain Increment Based Stress' for Steady-State problem

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17326] [Q&A Modules: Solid mechanics] [answered]
Hi MOOSE users,

I have question regarding 'Compute Strain Increment Based Stress' for steady-state simulation

![Screenshot from 2021-03-16 10-27-52](https://user-images.githubusercontent.com/79562439/111256008-3fce3280-866c-11eb-84b4-03df731c2d7e.png)

ultimately, I want my appliccation to work as follow, where I can change the value of C and strain at any particular block

![MicrosoftTeams-image](https://user-images.githubusercontent.com/79562439/111256545-6e98d880-866d-11eb-936c-7ae5f617e5ed.png)

So, the question is how do I define stress_old, strain_mech, strain_mech_old from an output file from the previous steady-state step?

Here in I attached the previous step output file generator (grav_redo.i)
and currect step .i file with Compute Strain Increment Based Stress (grav_redo_later.i)

[tensor mech elastic equli.zip](https://github.com/idaholab/moose/files/6146128/tensor.mech.elastic.equli.zip)

Thank you for your time.

Kind regards,
Traiwit

**Comment by ngrilli** [comment, marked as answer]
Dear @Traiwit 

In general, every MaterialProperty defined in the code can be initialised from a previous file.
For instance we add the following in the [AuxVariables] block of the restart input file to initialise a stress component:

  [./stress_xx]
          order = CONSTANT
          family = MONOMIAL
          initial_from_file_var = stress_xx
          initial_from_file_timestep = 620
  [../]

We have to do this for all the components.
If the above block is used, the standard initialisation in the material object is skipped.
You will need to do that for your: stress_old, strain_mech, strain_mech_old

So in general, if you want to restart a material model, you need to define in the initial input file
all the material properties that must keep the memory of the previous simulation (both in AuxVariables and AuxKernels),
then in the restart file, you define them again, both in AuxVariables and AuxKernels,
and you have to initialise them as in the example above. You need to do this for all components
of all MaterialProperty that must keep memory in your simulation.

Some MaterialProperty are like history variables and need to be initialised, others may not be history variables.
This you need to determine yourself.
Don't forget to check parent classes of your material objects, as they may contain some of the MaterialProperty
that must be initialised.

Trust this helps.

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by Traiwit** [reply]
Hi @ngrilli,

First of all, thank you for always helping me out, I really appreciate your help!

I will give it a go and keep you updated.


Kind regards,
Traiwit

**Comment by Traiwit** [comment]
Sidenote: I wish there are more examples of the steady-state problem, as transient doesn't suit my objective.

It can be easily done using for-loop in python/matlab, but solving tensor-mechanical problems there is also difficult

**Comment by sapitts** [comment]
You might also consider looking into the restart and recovery capability for MOOSE simulations. More information [here](https://mooseframework.inl.gov/application_usage/restart_recover.html). I've used the restart capability successfully in quasi-static / transient simulations, although I have never tried it in a steady state tensor mechanics simulation, and the option may be of use here too

**Reply by Traiwit** [reply]
Hi @sapitts, yes I am using the 'restart' for the Variable (disp_x,y,z) but it cannot call the higher-order Auxvariable for some reason, Auxvariable constant/monomial can be imported with no issue using `initial_from_file_var`.

**Reply by sapitts** [reply]
@aeslaughter is there a MOOSE developer that could help with this question about restart with higher order monomial variables?

**Reply by aeslaughter** [reply]
@loganharbour @friedmud Do either of you know how to help?
